reframe/core/pipeline.py

```diff
@@ -808,37 +801,29 @@ def __new__(cls, *args, _rfm_use_params=False, **kwargs):
                         os.path.dirname(inspect.getfile(cls))
                     )
 
+        # Prepare initialization of test defaults
+        obj.__deferred_rfm_init = obj.__rfm_init__(*args,
+                                                   name=cls.__qualname__,
+                                                   prefix=prefix, **kwargs)
```

Comment:
Perhaps a little explanation why we need to defer `__rfm_init__` here could help. If I understand correctly, it's because `__rfm_init__` takes arguments whereas hooks do not and we want to execute it as a hook.

Reply:
With this PR we move the injection of the variables and the parameters to the `__call__` of the metaclass, which means that these are only available after this `__new__` has returned, and `__rfm_init__` needs some of the variables and parameters to be present.